### PR TITLE
MNT: Remove cmap_d colormap access

### PR DESCRIPTION
**Pull request by greglucas**
## PR Summary

This is the removal of the deprecated `mpl.cm.cmap_d` dictionary. This is related to #20853, but does not close that as none of the `mpl.cm` globals are deprecated here.

Note that rather than adding pop() to the Registry, I went with the `unregister()` that was already on `cm.unregister_cmap()`. I'm not sure if we'd rather it behave like a dictionary with item access/removal that way, or with the register/unregister methods. I don't have a major preference one way or the other.

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/matplotlib/cm.py [hunk 1]
```diff
@@ -177,23 +136,67 @@ def register(self, cmap, *, name=None, force=False):
...
+        if name in self:
+            if not force and name in self._builtin_cmaps:
+                msg = f"Trying to re-register the builtin cmap {name!r}."
+                raise ValueError(msg)
+            elif not force:
+                raise ValueError(
+                    f'A colormap named "{name}" is already registered.')
+            else:
+                _api.warn_external(f"Trying to register the cmap {name!r} "
+                                   "which already exists.")
```
Comment (timhoffm) [inline lib/matplotlib/cm.py:153]: Would reorder to
```suggestion
            if force:
                _api.warn_external(f"Trying to register the cmap {name!r} "
                                   "which already exists.")
            else:
                raise ValueError(
                    f"Trying to re-register the builtin cmap {name!r}."
                    if name in self._builtin_cmaps else
                    f'A colormap named "{name}" is already registered.')
```

File: lib/matplotlib/cm.py [hunk 2]
```diff
@@ -177,23 +136,67 @@ def register(self, cmap, *, name=None, force=False):
...
 
+        self._cmaps[name] = cmap.copy()
+
+    def unregister(self, name):
+        """
+        Remove a colormap from the registry.
+
+        You may not remove built-in colormaps.
```
Comment (timhoffm) [inline lib/matplotlib/cm.py:161]:
```suggestion
        You cannot remove built-in colormaps.
```
Not a native speaker: Is that more clear?

File: lib/matplotlib/cm.py [hunk 3]
```diff
@@ -177,23 +136,67 @@ def register(self, cmap, *, name=None, force=False):
...
+            The ColormapRegistry is a shared namespace that may be used
+            by multiple packages. Use `unregister` only if you know you
+            have registered that name before. In particular, do not
+            unregister just in case to clean the name before registering a
+            new colormap.
+
+        Parameters
+        ----------
+        name : str
+            The name of the colormap to be removed
```
Comment (timhoffm) [inline lib/matplotlib/cm.py:177]:
```suggestion
            The name of the colormap to be removed.
```
Following numpydoc conventions, also ellipses are terminated with a period.

File: lib/matplotlib/cm.py [hunk 4]
```diff
@@ -177,23 +136,67 @@ def register(self, cmap, *, name=None, force=False):
...
+
+        Raises
+        ------
+        ValueError
+            If you try to remove a default built-in colormap.
+        """
+        if name in self._builtin_cmaps:
+            raise ValueError(f"cannot unregister {name!r} which is a builtin "
+                             "colormap.")
+        return self._cmaps.pop(name, None)
```
Comment (timhoffm) [inline lib/matplotlib/cm.py:192]: I know this has been the interface before. But returning a colormap seems quite pointless. I cannot imagine a practical scenario in which I unregister and keep hold of the colormap to do something with it. Returning a bool seems more clear. Since we introduce this as a new API, we're free to make that change (but would need two extra lines in the original `unregister` to keep that backward-compatible).

File: lib/matplotlib/cm.py [hunk 5]
```diff
@@ -177,23 +136,67 @@ def register(self, cmap, *, name=None, force=False):
...
-globals().update(_cmap_registry)
-# This is no longer considered public API
-cmap_d = _DeprecatedCmapDictWrapper(_cmap_registry)
-__builtin_cmaps = tuple(_cmap_registry)
 
 # public access to the colormaps should be via `matplotlib.colormaps`. For now,
 # we still create the registry here, but that should stay an implementation
 # detail.
-_colormaps = ColormapRegistry(_cmap_registry)
+_colormaps = _cmap_registry = ColormapRegistry(_gen_cmap_registry())
```
Comment (timhoffm) [inline lib/matplotlib/cm.py:198]:
```suggestion
_colormaps = ColormapRegistry(_gen_cmap_registry())
```
we had the two names before because for historic reasons and against intuition `_cmap_registry` was a dict whereas `_colormaps` was the registry. If we can get away without the dict, we don't need that name anymore.

**Comment by timhoffm** [comment]
> Note that rather than adding pop() to the Registry, I went with the unregister() that was already on cm.unregister_cmap(). I'm not sure if we'd rather it behave like a dictionary with item access/removal that way, or with the register/unregister methods. I don't have a major preference one way or the other.

`Colormaps` does intentionally not take on the full `MutableMapping` API. Register/unregister are similar to insertion/deletion but follow special rules, so we don't want to extend the mapping analogy to the modification API.

File: lib/matplotlib/cm.py (hunk 1 above)
Comment (greglucas) [reply lib/matplotlib/cm.py]: Thanks! This feels a bit awkward to me in general... I'm actually wondering if want to outright refuse to override a builtin? Right now we could use `force=True` and replace the builtin, but in unregister we don't allow that removal.

File: lib/matplotlib/cm.py (hunk 4 above)
Comment (greglucas) [reply lib/matplotlib/cm.py]: I agree, you can call `colormaps.get()` right before unregistering. I actually took it one step farther and said we don't even need a return at all. Happy to add it in if you think it is valuable, but seems even cleaner to avoid any return value.

File: lib/matplotlib/cm.py (hunk 1 above)
Comment (timhoffm) [reply lib/matplotlib/cm.py]: Good point. I was just cleaning up the structure, not considering something else. But I agree: We should not allow overriding builtins. Please change.

File: lib/matplotlib/cm.py (hunk 4 above)
Comment (timhoffm) [reply lib/matplotlib/cm.py]: Further simplification is good. :+1:

File: lib/matplotlib/cm.py (hunk 1 above)
Comment (greglucas) [reply lib/matplotlib/cm.py]: Oops, I forgot to look at our current function. Unfortunately `cm.register_cmap()` doesn't use `force`, it uses `override_builtin`, even more explicitly allowing this behavior. So, I'm not sure how we want to handle this. I really don't think people should be changing a colormap in their underlying library and then re-registering it as the same name, but apparently there are a few (~2) cases that do that :(
https://github.com/search?q=override_builtin&type=issues

Comment (timhoffm) [reply lib/matplotlib/cm.py]: We have `override_builtin` for compatibility (https://github.com/matplotlib/matplotlib/pull/18503#issuecomment-890600347) and at least for now cannot get rid of it. However, I don't want to have that in the new ColormapRegistry API. Since we're here switching to having the registry as the underlying structure, we have to feed that parameter in privately. Either using a private parameter

```
def register_cmap(...):
    ...
    _colormaps.register(cmap, name=name, force=True, _override_builtin=override_builtin)
```

or using a private flag (with uglier code, but slight peference since it does not show up in the docs):

```
def register_cmap(...):
    ...
    _colormaps._allow_override_builtin = override_builtin
    _colormaps.register(cmap, name=name, force=True)
    _colormaps._allow_override_builtin = False
```

File: doc/api/next_api_changes/deprecations/22298-GL.rst [hunk 6]
```diff
@@ -0,0 +1,5 @@
+``cmap_d`` removal
+~~~~~~~~~~~~~~~~~~
+
+The deprecated ``matplotlib.cm.cmap_d`` access to global colormaps
+has been removed.
```
Comment (timhoffm) [inline doc/api/next_api_changes/deprecations/22298-GL.rst:5]:
```suggestion
has been removed. Use the `.ColormapRegistry` ``matplotlib.colormaps`` instead.
```

File: lib/matplotlib/cm.py [hunk 7]
```diff
@@ -239,21 +235,12 @@ def register_cmap(name=None, cmap=None, *, override_builtin=False):
...
-            _api.warn_external(msg)
-
-    if not isinstance(cmap, colors.Colormap):
-        raise ValueError("You must pass a Colormap instance. "
-                         f"You passed {cmap} a {type(cmap)} object.")
-
-    cmap._global = True
-    _cmap_registry[name] = cmap
-    return
+    # override_builtin is allowed here for backward compatbility
```
Comment (timhoffm) [inline lib/matplotlib/cm.py:238]: IMHO we should deprecate overwriting builtin colormaps. But that can be a separate PR.

**Review by timhoffm — APPROVED** [review]

File: lib/matplotlib/cm.py (hunk 7 above)
Comment (greglucas) [reply lib/matplotlib/cm.py:238]: I agree with deprecating that functionality and saving it for a separate PR :)

**Comment by greglucas** [comment]
Ping for a second review on this deprecation removal.

**Review by tacaswell — APPROVED** [review]
I pushed an additional commit with a new test.   @greglucas or anyone can merge on CI green, but would like if at least one more person looked at the test I added.

**Comment by neutrinoceros** [comment]
Hi ! This breaks yt since it is using `cmap_d` internally. I see that it was considered deprecated but we never caught that despite trying to treat warnings as errors in CI. I'm not asking for it to be reverted, I just wanted to raise awareness that maybe the deprecation warning wasn't working as intended.

**Comment by timhoffm** [comment]
Thanks for the report. For the deprecation period `cmap_d` has been a proxy to the original dict, which should warn upon access. Maybe we forgot to issue a warning for one case? How are you using `cmap_d`?

**Comment by timhoffm** [comment]
Edit: Ah, you have gated against the new version, but are trying to get the registry via a private attribute (which we have changed here). 

https://github.com/yt-project/yt/blob/6270f345039f39d25eff8db3cbda7739dbb99f3a/yt/visualization/color_maps.py#L91

You should instead use `matplotlib.colormaps` or `pyplot.colormaps`.

Note also, that you have `plt.colormaps()` twice, which returns a list of colormaps. This is continued to be supported, but the recommended way is to simply use `plt.colormaps` here as well, since `plt.colormaps` is now the registry/container.

**Comment by neutrinoceros** [comment]
Thanks @timhoffm, that's gold ! you're welcome to review my fix for this if you want to https://github.com/yt-project/yt/pull/3912
and I suggest we move the discussion there if needed. Apologies for the noise everyone !

**Comment by greglucas** [comment]
Thanks for testing the upcoming changes so fast, @neutrinoceros! Also an FYI since it looks like you're building from MPL and numpy source, we do upload main-branch wheels to: https://anaconda.org/scipy-wheels-nightly if it would be helpful to install from the binary instead.
